miniapps/autodiff/example.hpp

```diff
@@ -0,0 +1,877 @@
+      vparam[1] = 1e-8; // default epsilon
+      vparam[2] = 1.0;  // default load
+
+   }
+
+   pLaplaceAD(Coefficient &pp_, Coefficient &q, Coefficient &ld_)
+      : pp(&pp_), coeff(&q), load(&ld_)
+   {
+      vparam.SetSize(3);	   
+      vparam[0] = 2.0;  // default power
+      vparam[1] = 1e-8; // default epsilon
+      vparam[2] = 1.0;  // default load
+   }
+
+   virtual ~pLaplaceAD() {}
+
+   virtual double GetElementEnergy(const FiniteElement &el,
+                                   ElementTransformation &trans,
+                                   const Vector &elfun) override
+   {
+      double energy = 0.0;
+      const int ndof = el.GetDof();
+      const int ndim = el.GetDim();
+      const int spaceDim = trans.GetSpaceDim();
+      bool square = (ndim == spaceDim);
+      int order = 2 * el.GetOrder() + trans.OrderGrad(&el);
+      const IntegrationRule &ir(IntRules.Get(el.GetGeomType(), order));
+
+
+      Vector shapef(ndof);
+      // derivatives in isoparametric coordinates
+      DenseMatrix dshape_iso(ndof, ndim);
+      // derivatives in physical space
+      DenseMatrix dshape_xyz(ndof, spaceDim);
+      Vector grad(spaceDim);
+
+      Vector uu(4);     //[diff_x,diff_y,diff_z,u]
+
+      uu = 0.0;
+
+      // Calculates the functional/energy at an integration point.
+      MyQFunctor<double,Vector,Vector,4,3> qfunc;
+
+      double w;
+      double detJ;
+
+      for (int i = 0; i < ir.GetNPoints(); i++)
+      {
+         const IntegrationPoint &ip = ir.IntPoint(i);
+         trans.SetIntPoint(&ip);
+         w = trans.Weight();
+         detJ = (square ? w : w * w);
+         w = ip.weight * w;
+
+         el.CalcDShape(ip, dshape_iso);
+         el.CalcShape(ip, shapef);
+         // AdjugateJacobian = / adj(J),         if J is square
+         //                    \ adj(J^t.J).J^t, otherwise
+         Mult(dshape_iso, trans.AdjugateJacobian(), dshape_xyz);
+         // dshape_xyz should be divided by detJ for obtaining the real value
+         // calculate the gradient
+         dshape_xyz.MultTranspose(elfun, grad);
+
+         // set the power
+         if (pp != nullptr)
+         {
+            vparam[0] = pp->Eval(trans, ip);
+         }
+
+         // set the coefficient ensuring positiveness of the tangent matrix
+         if (coeff != nullptr)
+         {
+            vparam[1] = coeff->Eval(trans, ip);
+         }
+         // add the contribution from the load
+         if (load != nullptr)
+         {
+            vparam[2] = load->Eval(trans, ip);
+         }
+         // fill the values of vector uu
+         for (int jj = 0; jj < spaceDim; jj++)
+         {
+            uu[jj] = grad[jj] / detJ;
+         }
+         uu[3] = shapef * elfun;
+         // the energy is taken directly from the templated function
+         energy = energy + w * qfunc(vparam,uu);
+      }
+      return energy;
+   }
+
+   virtual void AssembleElementVector(const FiniteElement &el,
```

Comment:
Yes, we can do it, however, it will be valid only for a single state field. I still haven't figured out a good interface for mixed problems. It is on the to-do list.